rbd: add support for bigger size restore/clone PVC

Madhu-1 opened the pull request:
This PR allows the user to create the bigger PVC from the PVC and restore a snapshot to a bigger size PVC. and did require refactoring to avoid linter issues.


fixes: #1566
fixes: #1438

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
csi: remove size check when creating volume

remove the  bigger size validation when
creating a volume from a snapshot or when
creation a clone from a volume as we resized
the volume after cloning.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: resize the volume created from snapshot

If the requested volume size is greater than
the snapshot size, resize the cloned volume
after creating a clone from a snapshot.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cleanup: add cleanup helper for incorrect thick volume

added a new helper function called cleanupThickClone
to cleanup the snapshot and clone if the thick
provisioning is not fully completed.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: resize the volume created from volume

If the requested volume size is greater than
the parent volume size, resize the cloned volume
after creating a final clone from a parent volume.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: update size for rbdSnapshot struct

we need actual size of the rbdVolume
created for the snapshot, as we are not
storing the size of the snapshot in OMAP
we need to fetch the size from ceph cluster
and update the same on rbdSnapshot  struct.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: consider ErrImageNotFound during DeleteSnapshot

added a check to consider ErrImageNotFound error
during DeleteSnapshot operation, if the error
is ErrImageNotFound we need to ensure that image
is removed from the trash and also the rados
OMAP data is removed.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: expand filesystem during NodeStageVolume

If the volume with a bigger size is created
from a snapshot or from another volume we
need to exapand the filesystem also in the
csidriver as nodeExpand request is not triggered
for this one, During NodeStageVolume we can
expand the filesystem by checking filesystem
needs expansion or not.

If its a encrypted device, check the device
size of rbd device and the LUKS device if required
the device will be expanded before
expanding the filesystem.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
e2e: add e2e for bigger size restore

added e2e for below cases

Normal PVC snapshot restore to a bigger
size PVC (without encryption)

* Filesystem pvc restore to a bigger size
* Block pvc restore to a bigger size

Encrypted PVC snapshot restore to a bigger
size PVC

* Filesystem pvc restore to a bigger size
* Block pvc restore to a bigger size

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
e2e: add e2e for bigger size clone

added e2e for below cases

Normal PVC clone to a bigger
size PVC (without encryption)

* Filesystem pvc clone to a bigger size
* Block pvc clone to a bigger size

Encrypted PVC clone to a bigger
size PVC

* Filesystem pvc clone to a bigger size
* Block pvc clone to a bigger size

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1397,6 +1394,33 @@ func (cs *ControllerServer) DeleteSnapshot(
+	// cleanup the image from trash if the error is image not found.
+	err = rbdVol.ensureImageCleanup(ctx)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to delete rbd image: %s/%s with error: %v", rbdVol.Pool, rbdVol.VolName, err)
```
while moving the code around, you are allowed to clean logging functions up too, just log `rbdVol` as `%q` so that the standard formatting of the image name is used (including optional rados namespace).

Same counts for the below logging of `rbdSnap`.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -588,6 +596,14 @@ func (cs *ControllerServer) createVolumeFromSnapshot(
 
+	// resize the volume if the size is different
+	if rbdSnap.SizeBytes != rbdVol.VolSize {
+		err = rbdVol.resize(rbdVol.VolSize)
```
Reducing size is dangerous, that should not be done.

Consider placing these resize checks/actions in a helper function:

```go
err = rbdVol.Expand(rbdSnap.SizeBytes)
...

// Expand resizes the rbdImage to the given size, in case the rbdImage is smaller.
func (ri *rbdImage) Expand(size int) error {
    if ri.VolSize >= size {
         return nil
    }

    return ri.resize(size)
}
```

Madhu-1 replied:
the smaller size is not allowed for now in kubernetes  (CO). will send a follow-up patch to introduce this check for cephcsi (for both cephfs and rbd) to disallow that operation as it might cause some problems like data loss etc.

nixpanic replied:
Indeed **NEVER** reduce the size of an RBD volume.

Changing quotas for CephFS would be acceptable, in case the current consumption does not hit the limit.

Madhu-1 replied:
https://github.com/ceph/ceph-csi/issues/2718

Madhu-1 replied:
will do some testing for cephfs how it works the reducing size of a bigger size subvolume and will send follow-up for this one. opened #2718 for tracking it.

nixpanic replied:
I still would like to see the check for resizing in a helper function. This if+resize block is repeated a lot.

Madhu-1 replied:
moved to a helper function

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -435,24 +435,13 @@ func (ns *NodeServer) stageTransaction(
-		var ok bool
-		resizer := mount.NewResizeFs(utilexec.New())
-		ok, err = resizer.NeedResize(devicePath, stagingTargetPath)
+	// As we are suppoting the restore of a volume to a bigger size and
```
typo `suppoting`

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -463,6 +452,46 @@ func (ns *NodeServer) stageTransaction(
+		}
+		// Use mapper device path for fs resize
+		devicePath = mapperPath
+	}
```
empty lines between sections of code make things more readable, please use that to please the eyes of your fellow developers

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -463,6 +452,46 @@ func (ns *NodeServer) stageTransaction(
+			stagingTargetPath,
+			err)
+	}
+	if ok {
```
In most flows we return early in case there is an error or nothing needs to be done. In such a coding style checking for `!ok` on this line and returning fits better. It also reduces the indention of the `resizer.Resize()` call, making the flow a little simpler to follow when just glancing over the code.

nixpanic commented on e2e/rbd.go:
```diff
@@ -3383,6 +3385,114 @@ var _ = Describe("RBD", func() {
 				}
 			})
 
+			By("restore snapshot to a bigger size PVC", func() {
```
what is the benefit of these two nested `By()` calls?

Madhu-1 replied:
Having nested `by` helps to break larger testing into smaller pieces for better readability.

nixpanic reviewed (changes requested):
Mostly some small coding improvements to make understanding and maintaining the code a little easier. Nothing else outstanding, as far as I am concerned.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1053,9 +1053,33 @@ func genSnapFromSnapID(
 
+// updateSnapshotDetails will copies the details from the rbdVolume to the
+// rbdSnapshot. example copying size from rbdVolume to rbdSnapshot.
+func updateSnapshotDetails(rbdSnap *rbdSnapshot) error {
```
Not sure how I missed this in the preview review round...

Can `getImageInfo()` not be a function for `rbdImage` instead? That would make things much simpler.

nixpanic replied:
hmm, this has not been addressed yet?

Madhu-1 replied:
unfortunately, this is still needed as rbdImage is created with RbdSnapName and I moved the size to rbdImage struct.

Madhu-1 commented:
@pkalever PTAL mainly for encrypted PVC expansion.

Madhu-1 commented:
something is wrong. marking it as draft as i need to test all flows manuallly

Madhu-1 committed:
rbd: add RequestedVolSize to rbdVolume struct

when doing the internal operation to get the
latest details the rbd image size is also getting
updated and this will update the volume size also
without actual requested size we cannot do the
resize operation for bigger clones. This commit
adds a new field called RequestedVolSize to rbdVolume
struct to hold the user requested size.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1397,6 +1394,39 @@ func (cs *ControllerServer) DeleteSnapshot(
+	}
+
+	// cleanup the image from trash if the error is image not found.
+	err = rbdVol.ensureImageCleanup(ctx)
```
maybe make `ensureImageCleanup()` a function  of `rbdImage` so that you can call it on the `rbdSnap` directly?

Madhu-1 replied:
This needs more changes than expected as i need to update multiple functions which am not touching in this PR. will revisit when working on the codebase cleanup.

nixpanic replied:
ok, thanks!

nixpanic commented on internal/rbd/clone.go:
```diff
@@ -181,6 +181,16 @@ func (rv *rbdVolume) createCloneFromImage(ctx context.Context, parentVol *rbdVol
 	}
 
+	// resize the image to the requested size is greater than the current size
+	if rv.VolSize > parentVol.VolSize {
```
repetitive code blocks, really should be done with a helper function

Madhu-1 replied:
should be fixed now

Madhu-1 committed:
rbd: remove json tag from rbdVolume struct

as we are no longer supporting the v1.x
version of cephcsi. removing the json tag
used to store rbd volume details in configmap.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: move VolSize to rbdImage struct

move the Volsize to the rbdImage struct
as size is more applicable for rbdImage
as rbdImage is used for both rbdVolume
and rbdSnapshot.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: remove SizeBytes from rbdSnapshot struct

as we are moving the VolSize to rbdImage struct
we should reuse the same instead of maintaining
one more field in rbdSnapshot struct.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

humblec commented on e2e/rbd.go:
```diff
@@ -3338,6 +3340,199 @@ var _ = Describe("RBD", func() {
+				})
+			})
+
+			By("clone PVC to a bigger size PVC", func() {
```
Instead of adding a new test can we add one more clone case to the "exisitng" clone test ?

Madhu-1 replied:
all clone tests need not go through the resize tests adding a separate test so that this can be reused for cephfs.

humblec commented on e2e/rbd.go:
```diff
@@ -3338,6 +3340,199 @@ var _ = Describe("RBD", func() {
+			})
+
+			By("clone PVC to a bigger size PVC", func() {
+				By("clone PVC to bigger size pvc", func() {
```
why 2 `By`s sequentially here?

Madhu-1 replied:
https://github.com/ceph/ceph-csi/pull/2716#discussion_r770352888 already discussed here

Madhu-1 committed:
rbd: correct logging for clone

log the rbdVolume and the rbdSnapshot
after creating the clone from snapshot.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: move core fields to rbdImage struct

moved ParentName, ParentPool and ImageFeatureSet
fields to the rbdImage struct as these are the
first citizens on the rbdImage.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: get image details after creating clone

after creating the clone get the current
image details like size, creationTime,
imageFeatures etc from the ceph cluster.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: remove extra getImageInfo API call

as getImageInfo is already called inside
cloneRbdImageFromSnapshot function right
after creating the clone. remove the extra
API call to get the details again.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: remove extra logging from createBackingImage

we are already logging the rbd image details
and the snapshot details after creating the
clone.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: correct logging in createBackingImage

after creating the rbd image log the image
details corresponding for the request along
with the request name.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -107,6 +107,9 @@ type rbdImage struct {
 	// identifying this rbd image
 	VolID string `json:"volID"`
 
+	// VolSize is the size of the RBD image backing this rbdVolume.
```
nit: this is now part of `rbdImage`, not `rbdVolume`

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1001,7 +1001,7 @@ func (cs *ControllerServer) CreateSnapshot(
 	}
 	rbdSnap.RbdImageName = rbdVol.RbdImageName
-	rbdSnap.SizeBytes = rbdVol.VolSize
+	rbdSnap.VolSize = rbdVol.VolSize
```
this commit has a typo in the commit message (maintaing).

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1703,6 +1703,16 @@ func cleanupRBDImageMetadataStash(metaDataPath string) error {
 
+// expand checks if the requestedVolume size and the existing image size both
+// are same. If they are same, it returns nil else it resizes the image.
+func (rv *rbdVolume) expand() error {
```
I was expecting to see `expand()` to have a `size` parameter... I guess this works too.

Madhu-1 replied:
instead of passing one more input using struct fields internally

Madhu-1 replied:
instead of passing one more input using struct fields internally

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -427,32 +427,93 @@ func (ns *NodeServer) stageTransaction(
+
+	resizer := mount.NewResizeFs(utilexec.New())
+
+	if transaction.isEncrypted {
```
a helper like `resizeEncryptedDevice()` might be good to have?

nixpanic commented on e2e/clone.go:
```diff
@@ -0,0 +1,84 @@
+package e2e
```
probably should add a copyright/license header?

Madhu-1 replied:
we dont have copyright/license for the e2e tests

nixpanic replied:
but we should?

Madhu-1 replied:
will add to all the e2e files in follow-up.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -127,6 +127,12 @@ type rbdImage struct {
+	ParentName string
+	// Parent Pool is the pool that contains the parent image.
+	ParentPool      string
+	ImageFeatureSet librbd.FeatureSet
```
why export `ImageFeatureSet`?

Madhu-1 replied:
without this one, the golint was failing and most of the members of rbdImage are exported and it's for a reason?

nixpanic replied:
I hope that in the future we do not need to have the members exported, as that invites to not have a nice `rbdImage` API. It is fine for now.

humblec commented on e2e/rbd.go:
```diff
@@ -3338,6 +3340,199 @@ var _ = Describe("RBD", func() {
+					}
+				})
+
+				By("restore snapshot to bigger size encrypted PVC with VaultKMS", func() {
```
`s/VaultKMS/Vault KMS` .

Madhu-1 replied:
this is a copy-paste from `By("create a PVC and bind it to an app with encrypted RBD volume with VaultKMS", func() {`

Madhu-1 replied:
`By("Validate PVC restore from vaultKMS to vaultTenantSAKMS", func() {`

humblec replied:
ok.. then we can keep the same format

humblec commented on examples/rbd/pod-block-restore.yaml:
```diff
@@ -0,0 +1,17 @@
+---
```
instead of adding new yaml here, can we stick or use exisitng one `pod-with-raw-block-volume` ?

Madhu-1 replied:
as new PVC is added keeping a different pod name

humblec replied:
yeah, thats what I was trying to avoid. As no changes in the PVC yamls. so I am checkign why we need new PVC and POD. Cant we use existing ones?

Madhu-1 replied:
not possible as we don't have the existing yamls in place, i don't see it as a major concern/blocker for this one

humblec commented on examples/rbd/pvc-block-restore.yaml:
```diff
@@ -0,0 +1,17 @@
+---
```
same case here `rbd-pvc-restore`

Madhu-1 replied:
rbd-pvc-restore is a filesystem PVC this is for BlockPVC

humblec replied:
it was a typo-- , I mean we have existing FS mode and Block Mode PVC yamls which can be reused..

Madhu-1 replied:
there is no existing block PVC restore.

nixpanic reviewed (dismissed):
No real concerns. It would be good to have the copyright/licence header in new e2e files, but that is not a blocker.

Madhu-1 committed:
e2e: add Copyright header of e2e files

added Copyright header for files inside
the e2e folder.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
@nixpanic added Copyright for all files inside the e2e folder.

Madhu-1 commented:
@humblec PTAL there  is a requested change from you

humblec commented:
Adding 3.5 milestone here for the tracker.

humblec commented:
mostly looks good.. few minor comments.

Madhu-1 commented:
@humblec PTAL

humblec commented:
As discussed this looks good to me .. thanks . :+1:

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
/retest all

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
/retest ci/centos/k8s-e2e-external-storage/1.21
